tests/cpp/test_pointwise.cpp

```diff
@@ -773,4 +783,112 @@ TEST_F(PointwiseTest, VectorizePadLoweringPermuted) {
   EXPECT_TRUE(found_vectorize);
   testValidate(&fusion, cg_outputs, aten_inputs, __LINE__, __FILE__);
 }
+
+TEST_F(PointwiseTest, DomainMapTestEg0) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  auto fusion = fusion_ptr.get();
+  FusionGuard fg(fusion);
+
+  TensorView* tv0 = makeContigTensor(2);
+  fusion->addInput(tv0);
+  auto tv1 = relu(tv0);
+  fusion->addOutput(tv1);
+  auto tv2 = broadcast(tv1, {false, true, false});
+  auto tv3 = expand(
+      tv2,
+      {tv2->axis(0)->extent(),
+       IrBuilder::create<Val>(4),
+       tv2->axis(2)->extent()});
+  auto tv4 = reshape(tv3, {2, 4, 3}, {2, 12});
+  fusion->addOutput(tv4);
+
+  DomainMapUnitTest domain_map(fusion);
+  // tv1 can't map to tv4, because the expanded ID participates in
+  // transformation
+  EXPECT_FALSE(domain_map.testOutputMapping(tv4, tv1));
+
+  // tv1 can map to tv3, because the missing ID is broadcast
+  EXPECT_TRUE(domain_map.testOutputMapping(tv3, tv1));
+
+  // tv4 can map to tv1
+  EXPECT_TRUE(domain_map.testOutputMapping(tv1, tv4));
+
+  // tv1 is not a valid reference
+  EXPECT_FALSE(domain_map.isValidReference(tv1));
+
+  // tv4 is a valid reference
+  EXPECT_TRUE(domain_map.isValidReference(tv4));
+}
+
+TEST_F(PointwiseTest, DomainMapTestEg1) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  auto fusion = fusion_ptr.get();
+  FusionGuard fg(fusion);
+
+  TensorView* tv0 = makeContigTensor(2);
+  fusion->addInput(tv0);
+  TensorView* tv1 = makeContigTensor(3);
+  fusion->addInput(tv1);
+  auto tv2 = reshape(tv0, {2, 4}, {8});
+  fusion->addOutput(tv2);
+
+  auto tv3 = broadcast(tv0, {true, false, false});
+  auto tv4 = add(tv1, tv3);
+  fusion->addOutput(tv4);
+
+  DomainMapUnitTest domain_map(fusion);
+  // tv2 can't map to tv4, because it misses tv4->axis(0)
+  EXPECT_FALSE(domain_map.testOutputMapping(tv4, tv2));
+
+  // tv2 can map to tv4
+  EXPECT_TRUE(domain_map.testOutputMapping(tv2, tv4));
+
+  // However, tv2 is not a valid reference, since it doesn't cover all input IDs
+  EXPECT_FALSE(domain_map.isValidReference(tv2));
+
+  // tv4 is a valid reference
+  EXPECT_TRUE(domain_map.isValidReference(tv4));
+}
+
+TEST_F(PointwiseTest, DomainMapFactory) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  auto fusion = fusion_ptr.get();
+  FusionGuard fg(fusion);
+
+  TensorView* tv0 = makeContigTensor(1);
+  fusion->addInput(tv0);
+  TensorView* tv1 = makeContigTensor(2);
+  fusion->addInput(tv1);
+
+  auto tv2 = broadcast(tv0, {true, true, false});
+  auto tv3 = add(tv2, tv1);
+  fusion->addOutput(tv3);
+
+  auto size_val = IrBuilder::create<Val>(4.0, DataType::Int);
+  auto one_val = IrBuilder::create<Val>(1, DataType::Int);
+  auto tv4 = ones({size_val, one_val, tv0->axis(0)->extent()}, DataType::Float);
+  auto tv5 = add(tv2, tv4);
+  fusion->addOutput(tv5);
+
```

Comment:
The expectation is simply a functional fusion.

The reason I added this is that I found factory methods causes an assert. i.e. we falsely assume one of the output to be a legit reference and try to schedule like that, which triggers an assert.

Reply:
How should this fusion be scheduled with which tensor as a reference?

Reply:
I'm not entirely sure what you mean by:
> The added check on the segmentation/reference_tv doesn't mean anything then

Reply:
by `added checks`, I was referring to the suggestion you made.

>  let's assert this is segmented to 2 fusions with each having t4 or t6 as its reference, respectively.

Reply:
Why will they be meaningless?

Reply:
My understanding is that, failing the assert doesn't signal any other actionable items, other than modifying the test to relax these checks.

I'm not saying it's meaningless, I'm just not clear on the engineering philosophy behind it.

Reply:
I'm sorry but I don't get it at all. This segmentation is what we expect from this fusion because there's no single reference tensor. The assertion here is to make sure the scheduler incorrectly tries to schedule the whole fusion without segmentation. The scheduler should also pick the tensors for each of the two segments as long as the logic of picking the reference works as intended.

We should not simply remove the assertion just because they fail, but we should understand why and find out how to fix them.

Please let me know if I'm missing anything.

Reply:
Let me rephrase what I was suggesting:
Instead of asserting `num_of_segment == 2`, I think we would prefer something like assert `num_of_segment <= 2`, just to ensure that we are not perf regression by creating more segments than the best we can do today.

I'm not saying that we should remove asserts to *hide* failures, I'm saying that `assert num_of_segment == 2` could be a false negative test in the future when our scheduler evolved so that we are capable of scheduling a single kernel for this fusion pattern.

Reply:
This fusion should be segmented as you mentioned in the beginning. Not segmenting is an error at this moment, so I don't think we should use `<=`. 

If the scheduler is improved in the future, this test would fail, and that's a clear signal that we would need to pay attention to this test. I'd happy to do the necessary update since that's a positive improvement, and I think the cost of the update is pretty cheap compared to the benefit of having the piece of mind that the segmentation is working as intended.

That's the basic philosophy of testing in nvFuser. That's why, for example, we do things like literal string comparison of generated code, which tends to yield a lot of noises but sometimes gives us important signals for detecting unintended behavior.

Reply:
Thanks, I see where the motivation is coming from. I'll update the test.